snuba/consumers/snapshot_worker.py

```diff
@@ -0,0 +1,76 @@
+import logging
+
+from confluent_kafka import Producer
+from typing import Any, Optional, Mapping
+
+from snuba.consumer import ConsumerWorker, KafkaMessageMetadata
+from snuba.stateful_consumer.control_protocol import TransactionData
+from snuba.datasets import Dataset
+from snuba.snapshots import SnapshotId
+
+logger = logging.getLogger('snuba.snapshot-consumer')
+
+
+class SnapshotAwareWorker(ConsumerWorker):
+    def __init__(self,
+        dataset: Dataset,
+        producer: Producer,
+        snapshot_id: SnapshotId,
+        transaction_data: TransactionData,
+        replacements_topic: Optional[str],
+        metrics: Optional[Any] = None,
+    ) -> None:
+        super(SnapshotAwareWorker, self).__init__(
+            dataset=dataset,
+            producer=producer,
+            replacements_topic=replacements_topic,
+            metrics=metrics,
+        )
+        self.__snapshot_id = snapshot_id
+        self.__transaction_data = transaction_data
+        self.__catching_up = True
+        self.__skipped_batch = []
+        self.__xip_list_applied = []
+        logger.debug("Starting snapshot aware worker for id %s", self.__snapshot_id)
+
+    def _process_message_impl(
+        self,
+        value: Mapping[str, Any],
+        metadata: KafkaMessageMetadata,
+    ):
+        if self.__catching_up:
+            xid = value.get("xid")
+            if xid:
+                if xid >= self.__transaction_data.xmax:
+                    logger.info(
+                        "Found xid(%s) >= xmax:(%s). Catch up phase is over",
+                        xid,
+                        self.__transaction_data.xmax,
+                    )
+                    logger.info(
+                        "Skipped %d transactions. Applied %d from xip_list",
+                        len(self.__skipped_batch),
```

Comment:
Are these actually transactions? Aren't they just statements/mutations? Or are they being grouped together somewhere else? If that's the case, aren't there going to be a lot of duplicates in these lists? Should they be something else like [`Counter`](https://docs.python.org/3/library/collections.html#collections.Counter) types so that you can tell the number of transactions and/or events?

Reply:
They are events. this is bad naming. Will change.
And you are right, I should group them to provide some meaningful number.